### ISSUE TITLE
How to insert null values ?

### ISSUE THREAD

**Issue by hansgru** [scriptella/scriptella-etl#16]
How to insert null values (e.g. in the case of db 2 db, so not using a CSV file).

I have something like:
```xml
    <query connection-id="db-in">
        select col1, col2, col3 from table1_in
        <script connection-id="db-out">
            insert into table1_out(col1, col2, col3)
            values($col1, $col2, $col3)
        </script>
    </query>
```

It seems to work, except for null values: in that case I'm getting an error:
```Driver exception: org.h2.jdbc.JdbcSQLException: Syntax error in SQL statement "```

Thank you.

**Comment by seansummers** [comment]
Does this work if you use a SQL prepared statement? You would have to change "$" to "?" in your script, as referenced in the [Scriptella ETL Reference Documentation: Expressions and Variables Substitution](http://scriptella.javaforge.com/reference/index.html#Expressions+and+Variables+Substitution).

**Comment by bhuesemann** [comment]
I think this is root in the same cause as mentioned here: http://scriptella.javaforge.com/faq.html#NULLS
It will work fine when using prepared statements like Sean suggested. At any time try to use prepared statements anyways.

**Comment by hansgru** [comment]
@seansummers and @bhuesemann 
Thanks you for your support :)

Indeed, using "?" is fixing the problem. 

I had this problem in the past, but I still do this mistake, especially because of copy and paste: e.g. when there are many columns and many tables.

What I'm really missing is some sort of "macro" to correctly expand all columns from select to the right place. E.g.:

```xml
   <query connection-id="db-in">
        select col1, col2, col3,...,coln from table1_in
        <script connection-id="db-out">
            insert into table1_out(#ALL_COLUMNS_FROM_RESULTSET)
            values(#ALL_COLUMNS_FROM_RESULTSET)
        </script>
    </query>
```

**Comment by seansummers** [comment]
Unfortunately, that work is a necessary evil of ETL (yay for humans!). Usually this is solved by a process that writes the etl.xml and then executes it, so you can consider the etl.xml file documentation of a specific process. I believe I once wrote a process that exported the columns to a text file and then read them back in for the insert, but it was brittle, confusing, and hard to implement so that the statement could be prepared for efficient execution -- much more work than knowing (now and later) what columns were involved and slowing down the process each run.

I vote #wontfix, as this is not the tool for that job ([reference](http://www.parseerror.com/blog/select-*-is-evil)).

**Comment by hansgru** [comment]
> .. much more work than knowing (now and later) what columns were involved and slowing down the process each run.

I would not want something with ```select * from table_in```, but the ```select``` statement to drive the ```insert```, e.g.
```sql
select col1 as col1_out, col2 as col2_out, ... coln as coln_out from table_in
```
since the column alias in the ```select``` can be used to match target to insert. This alone would already save allot of time and would avoid many bugs.

**Comment by bhuesemann** [comment]
I'm generating the scriptella jobs with scriptella itself (e.g. for simple staging layers). This works perfectly and reduces error sources.

**Comment by hansgru** [comment]
> I'm generating the scriptella jobs with scriptella itself (e.g. for simple staging layers). 

IMHO than something with Scriptella itself is broken :).

We separated the select SQLs to external files (since they are complicated and edited by DB specialists), but if we would have to generate Scriptella XML config files than we wouldn't not need the Scriptella overhead at all - we could simply do this directly in Java with some libs.

**Comment by bhuesemann** [comment]
How did you externalize the prepared inserts? I had some problems to do that because the parameters substitutation didn't work.

**Comment by hansgru** [comment]
> How did you externalize the prepared inserts?

Sorry but we did this only for the ```select``` statements because:
 - the Scriptella notation for inserts is not "SQL standard" so DB devs do not know it, nor do tools help there, nor can it be easily and in a generic way tested separately.
 - All the complicated logic is in Select statements, so that the insert is really only ```insert into ... values ...```
 - Our selects don't require parameters, but we use more stages of tables until the final inserts, because of cleanup/data quality reporting requirements.
 - Each Select SQL file has a header (comment) where the column names and types are enforced and with a different tool also checked. This way the DB devs work only with SQL files.

The only other problem we had with external SQL files was the "default" file encoding: we needed to enforce Scriptella to always use ```UTF-8``` by changing in ```bin/scriptella.bat``` the last line:
```bat
rem ---- run scriptella
%JAVACMD% -Dfile.encoding=UTF8 -cp %_SCRIPTELLA_CP% scriptella.tools.launcher.EtlLauncher %1 %2 %3 %4 %5 %6 %7 %8 %9 
```
otherwise Scriptella would use whatever platform was running on, so the operations were not consistent among different machines.
